### PR TITLE
[10.0] Port product_barcode_generator

### PR DESCRIPTION
**Pull request by JosDeGraeve**

### CODE REVIEW

File: product_barcode_generator/__init__.py [hunk 1]
```diff
@@ -1,7 +1,7 @@
 # -*- coding: utf-8 -*-
 ###############################################################################
```
Comment (pedrobaeza) [inline product_barcode_generator/__init__.py:2]: Shorten headers like in https://github.com/OCA/maintainer-tools/blob/master/template/module/__init__.py

**Comment by pedrobaeza** [comment]
Check Travis

File: product_barcode_generator/models/product.py [hunk 2]
```diff
@@ -30,67 +30,67 @@ def isodd(x):
 class ProductCategory(models.Model):
     _inherit = 'product.category'
 
-    ean_sequence_id = fields.Many2one('ir.sequence', string='Ean sequence')
+    barcode_sequence_id = fields.Many2one(comodel_name='ir.sequence',
+                                          string='Ean sequence')
```
Comment (Garamotte) [inline product_barcode_generator/models/product.py:34]: Please add the `oldname` attribute for renamed fields.

File: product_barcode_generator/__manifest__.py [hunk 3]
```diff
@@ -23,19 +23,19 @@
...
-        "data/ean_sequence.xml",
+        "data/barcode_sequence.xml",
         "views/res_company_view.xml",
         "views/product_view.xml",
         "views/sequence_view.xml",
     ],
-    'installable': False,
-    "active": False,
+    'installable': True,
+    "active": True,
```
Comment (Garamotte) [inline product_barcode_generator/__manifest__.py:40]: Remove the `active` key.

File: product_barcode_generator/models/product.py [hunk 4]
```diff
@@ -30,67 +30,67 @@ def isodd(x):
...
 
 
 class ProductProduct(models.Model):
```
Comment (Garamotte) [inline product_barcode_generator/models/product.py:37]: Rename the class to `ProductTemplate`.

File: product_barcode_generator/models/product.py [hunk 5]
```diff
@@ -30,67 +30,67 @@ def isodd(x):
...
-    _inherit = 'product.product'
+    _inherit = 'product.template'
+
+    @api.one
```
Comment (Garamotte) [inline product_barcode_generator/models/product.py:40]: There is no need to loop here.
Use `@api.multi` instead of `@api.one`.

File: product_barcode_generator/models/product.py [hunk 6]
```diff
@@ -30,67 +30,67 @@ def isodd(x):
...
+    def generate_barcode(self):
+        return self.product_variant_ids.generate_barcode()
+
 
-    ean13 = fields.Char(copy=False)
-    ean_sequence_id = fields.Many2one('ir.sequence', string='Ean sequence')
+class ProductProduct(models.Model):
+    _inherit = 'product.product'
 
     @api.model
```
Comment (Garamotte) [inline product_barcode_generator/models/product.py:48]: Change the decorator to `@api.multi` when you read values from fields.

File: product_barcode_generator/models/product.py [hunk 7]
```diff
@@ -30,67 +30,67 @@ def isodd(x):
...
-    _inherit = 'product.product'
+    _inherit = 'product.template'
+
+    @api.one
+    def generate_barcode(self):
+        return self.product_variant_ids.generate_barcode()
+
 
-    ean13 = fields.Char(copy=False)
-    ean_sequence_id = fields.Many2one('ir.sequence', string='Ean sequence')
```
Comment (Garamotte) [inline product_barcode_generator/models/product.py:40]: Why did you remove this field ?

File: product_barcode_generator/models/product.py [hunk 8]
```diff
@@ -30,67 +30,67 @@ def isodd(x):
...
-    def _get_ean_next_code(self, product):
-        sequence_obj = self.env['ir.sequence']
-        if product.ean_sequence_id:
-            ean = sequence_obj.next_by_id(product.ean_sequence_id.id)
-        elif product.categ_id.ean_sequence_id:
-            ean = sequence_obj.next_by_id(product.categ_id.ean_sequence_id.id)
-        elif product.company_id and product.company_id.ean_sequence_id:
-            ean = sequence_obj.next_by_id(
-                product.company_id.ean_sequence_id.id)
-        elif self.env.context.get('sequence_id', False):
```
Comment (Garamotte) [inline product_barcode_generator/models/product.py:52]: Why did you remove the context sequence_id lookup ?

File: product_barcode_generator/models/product.py [hunk 9]
```diff
@@ -30,67 +30,67 @@ def isodd(x):
...
                   "the 13 being used as a control digit"
                   "You will have to redefine the sequence or create a new one")
-                )
+            )
 
-        return ean
+        return barcode
 
-    def _get_ean_control_digit(self, code):
+    def _get_barcode_control_digit(self, barcode):
```
Comment (Garamotte) [inline product_barcode_generator/models/product.py:71]: This method is now available in the standard model `barcode.nomenclature` of the `barcodes` module, maybe you could use it ?
But this would require a new dependency...

File: product_barcode_generator/views/sequence_view.xml [hunk 10]
```diff
@@ -1,5 +1,5 @@
 <?xml version="1.0" encoding="utf-8"?>
-<openerp>
+<odoo>
     <data>
```
Comment (Garamotte) [inline product_barcode_generator/views/sequence_view.xml:3]: Remove the `<data>` tag.

File: product_barcode_generator/views/res_company_view.xml [hunk 11]
```diff
@@ -1,22 +1,23 @@
 <?xml version="1.0" encoding="utf-8"?>
-<openerp>
+<odoo>
     <data>
```
Comment (Garamotte) [inline product_barcode_generator/views/res_company_view.xml:3]: Remove the `<data>` tag.

File: product_barcode_generator/views/product_view.xml [hunk 12]
```diff
@@ -1,35 +1,52 @@
 <?xml version="1.0" encoding="utf-8"?>
-<openerp>
+<odoo>
     <data>
```
Comment (Garamotte) [inline product_barcode_generator/views/product_view.xml:3]: Remove the `<data>` tag.

**Review by Garamotte — CHANGES_REQUESTED** [review]

File: product_barcode_generator/data/barcode_sequence.xml [hunk 13]
```diff
@@ -0,0 +1,16 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo>
```
Comment (lasley) [inline product_barcode_generator/data/barcode_sequence.xml:2]: This can be simplified as `<odoo noupdate="1">` & can remove data tag

File: product_barcode_generator/models/product.py (hunk 5 above)
Comment (lasley) [reply product_barcode_generator/models/product.py:40]: Additionally, api.one is deprecated so we can't use it anymore 😦

File: product_barcode_generator/models/product.py [hunk 14]
```diff
@@ -30,67 +30,67 @@ def isodd(x):
...
-            ean = sequence_obj.next_by_id(product.ean_sequence_id.id)
-        elif product.categ_id.ean_sequence_id:
-            ean = sequence_obj.next_by_id(product.categ_id.ean_sequence_id.id)
-        elif product.company_id and product.company_id.ean_sequence_id:
-            ean = sequence_obj.next_by_id(
-                product.company_id.ean_sequence_id.id)
-        elif self.env.context.get('sequence_id', False):
-            ean = sequence_obj.next_by_id(self.env.context.get('sequence_id'))
+    def _get_barcode_next_code(self, ):
+        if self.categ_id.barcode_sequence_id:
```
Comment (lasley) [inline product_barcode_generator/models/product.py:50]: Looks like this expects a singleton, will need a `self.ensure_one()`

File: product_barcode_generator/models/product.py (hunk 9 above)
Comment (lasley) [reply product_barcode_generator/models/product.py:71]: I think the dependency is worth it TBH. Will allow for this to fit with additions to it like #60

File: product_barcode_generator/models/product.py [hunk 15]
```diff
@@ -30,67 +30,67 @@ def isodd(x):
...
+    def _generate_barcode_value(self):
+        barcode = self._get_barcode_next_code()
+        if not barcode:
             return None
-        key = self._get_ean_control_digit(ean)
-        ean13 = ean + key
-        return ean13
+        return barcode + self._get_barcode_control_digit(barcode)
 
     @api.one
```
Comment (lasley) [inline product_barcode_generator/models/product.py:88]: api.one is deprecated

**Review by lasley — COMMENTED** [review]
Thanks @JosDeGraeve - few comments inline

File: product_barcode_generator/models/product.py (hunk 2 above)
Comment (lasley) [reply product_barcode_generator/models/product.py:34]: We may also want to change the string attribute if we're abstracting the definition of the field for barcode vs just ean.

**Comment by lasley** [comment]
On second thought, I think [`barcodes_generator_product`](https://github.com/OCA/stock-logistics-barcode/tree/9.0/barcodes_generator_product) is the successor to this module

cc @legalsylvain

**Comment by legalsylvain** [comment]
indeed. @josdegraeve : Did you taken a look on this module ? It is based on barcode nomenclature core feature and allow sequenced or manual numerotation.

**Comment by lasley** [comment]
I went ahead and just upgraded them all for good measure, I was about to need 2 of the 3 anyways. #65, #66, #67

**Comment by lasley** [comment]
Closing, #66 supersedes because this module is deprecated & removed
